lspci: Improve CXL DVSEC decoding

jphaws opened the pull request:
I wanted to help #59 along since it's been open for awhile and I'm interested in following up with additional CXL DVSEC decoding. I addressed the accessing cached config data by type-casting pointers issues by replacing them with their respective `get_conf_xxx` and rebased against the current master. Other than that, everything remains as originally written by @bwidawsk.

pali commented on ls-ecaps.c:
```diff
@@ -701,7 +701,7 @@ cap_dvsec_cxl(struct device *d, int id, int where)
     return;
 
-  if (!config_fetch(d, where + PCI_CXL_CAP, 12))
+  if (!config_fetch(d, where + PCI_CXL_CAP, 0x38 - 0xa))
```
If you are changing _size_ by its calculation then I would suggest to use new/existing macros instead of harcoding magic numbers like `0x38` or `0xa`. There are:
```c
#define PCI_CXL_CAP		0x0a	/* CXL Capability Register */
#define PCI_CXL_DEV_CAP			0x0a	/* CXL Capability Register */
#define PCI_CXL_DEV_RANGE2_BASE_LO	0x34
```
which seems to be source of those magic numbers.

pali commented on ls-ecaps.c:
```diff
@@ -690,35 +690,49 @@ cap_rcec(struct device *d, int where)
-
-  if (!config_fetch(d, where + PCI_CXL_CAP, 0x38 - 0xa))
+  /* Legacy 1.1 revs aren't handled */
+  if (rev != 1)
```
Cannot be revision bigger than 1 in future with backward compatible capabilities? IIRC All PCI capabilities were always defined in backward compatible manner. So check for one specific revision seems to be unexpected.

jphaws replied:
Yes, you are correct. CXL capabilities are defined to be backwards compatible with future revisions. Checking against a specific revision of capabilities without also checking future revisions prevents forward compatibility with later revisions. I'll push an amended commit addressing this.

pali reviewed (commented):
Hello! I looked at this new version and below are my two comments:

pali commented on ls-ecaps.c:
```diff
@@ -690,30 +690,206 @@ cap_rcec(struct device *d, int where)
+  cxl_range(range_base, range_size, 2);
+}
+
+static void
+dvsec_cxl_port(struct device *d, int where, int rev)
+{
+  u16 w, m1, m2;
+  u8 b1, b2;
+
+  if (rev != 0)
+    return;
```
Should not be CXL port capabilities also forward compatible?

jphaws replied:
Yes, they should be. I'll take look back over revision checks throughout and make sure that the code is forward compatible.

pali commented on ls-ecaps.c:
```diff
@@ -690,30 +690,206 @@ cap_rcec(struct device *d, int where)
+  u64 range_base, range_size;
+  u16 w;
+
+  /* Legacy 1.1 revs aren't handled */
+  if (rev < 1)
     return;
```
How is mapped `int rev` to the revision tuple major.minor? Because this check reject rev==0 but comment say that 1.1. So seems like there is non-trivial mapping between integer and tuple major.minor (if it is really major.minor format).

jphaws replied:
From 8.1.3 of the CXL 3.0 specification, the DVSEC Revision ID is mapped as follows: 
`
DVSEC Revision ID 0h represents the structure as defined in the CXL 1.1 specification.
`
`
DVSEC Revision ID 1h represents the structure as defined in the CXL 2.0 specification.
`
`
DVSEC Revision ID 2h represents the structure as defined in the CXL 3.0 specification.
`

pali replied:
Thanks! Interesting mapping... and code with comment looks correct.

pali reviewed (commented):
Hello! Could you look at following two comments?

jphaws commented:
Addressed additional revision checks that would result in capability decoding not being forward compatible with future revisions. Also rebased separate commit that addressed `get_conf_xxx` into commits where `get_conf_xxx` was not used.

pali commented on ls-ecaps.c:
```diff
@@ -690,30 +690,206 @@ cap_rcec(struct device *d, int where)
+  u16 w, m1, m2;
+  u8 b1, b2;
+
+  if (rev < 0)
+    return;
```
It is possible that this condition is truth in some case?
Type of `rev` is signed integer but you fill it via `rev = BITS(get_conf_byte(d, where + 0x6), 0, 4);`
which is always non-negative number and of unsigned type `u8`.

jphaws replied:
It seems I was mistaken when I though further revisions had been made to this register and the one below. The method used to fill `rev`, as you mentioned, will never fill to anything less than 0. Since this condition will never be true and they are both currently on revision 0 (section 8.1.8 & section 8.1.5), I'm going to remove these revision checks as well.

pali replied:
Ok, now it looks better!

pali commented on ls-ecaps.c:
```diff
@@ -690,30 +690,206 @@ cap_rcec(struct device *d, int where)
+{
+  int i, j;
+
+  if (rev < 0)
+    return;
```
Same here.

pali reviewed (commented):
Just one suspicious thing:

jphaws commented:
@gollux is there anything else I can do to help this one along?

gollux commented:
I merged the patch series, but I had to clean it up a lot. Please see my commit in `master` and check it.

Also, I would appreciate if you could add a configuration dump of some device with CXL to `tests/`, so that I can test the code.

I noticed some of the capabilities are still not fully decoded. I can add it, but I do not want to do it without test data as I do not have access to any real CXL device :)

jphaws commented:
Just opened #115 to address a bug with your commit to `master`, otherwise it looks good to me. 

I have another series pending to decode the remaining capabilities that I'll rebase and clean up over the next few days before sending it out as well. I need to double check before publishing the configuration dump, but it shouldn't be a problem :)
